tests/python_package_test/test_engine.py

```diff
@@ -1434,25 +1431,220 @@ def decreasing_metric(preds, train_data):
 
         def constant_metric(preds, train_data):
             return ('constant_metric', 0.0, False)
-
         # test that all metrics are checked (default behaviour)
+        params = {
+            'objective': 'regression',
+            'metric': 'None',
+            'verbose': -1
+        }
         gbm = lgb.train(params, lgb_train, num_boost_round=20, valid_sets=[lgb_eval],
                         feval=lambda preds, train_data: [decreasing_metric(preds, train_data),
                                                          constant_metric(preds, train_data)],
                         early_stopping_rounds=5, verbose_eval=False)
         self.assertEqual(gbm.best_iteration, 1)
-
         # test that only the first metric is checked
+        try:
+            gbm = lgb.train(dict(params, first_metric_only=True), lgb_train,
+                            num_boost_round=20, valid_sets=[lgb_eval],
+                            feval=lambda preds, train_data: [decreasing_metric(preds, train_data),
+                                                             constant_metric(preds, train_data)],
+                            early_stopping_rounds=5, verbose_eval=False)
+            self.assertTrue(False, "LightGBMError should be raised.")
+        except lgb.basic.LightGBMError as e:
+            self.assertEqual(e.args[0], '`first_metric_only` and `feval` are not available at the same time.')
+        # test that various combination of metrics.
+        params = {
+            'boosting_type': 'gbdt',
+            'objective': 'regression',
+            'learning_rate': 0.5,
+            'num_leaves': 5,
+            'metric': 'l2',
+            'verbose': -1,
+            'seed': 123,
+            'bagging_seed': 234,
+            'feature_fraction_seed': 456
+        }
         gbm = lgb.train(dict(params, first_metric_only=True), lgb_train,
-                        num_boost_round=20, valid_sets=[lgb_eval],
+                        num_boost_round=300, valid_sets=[lgb_eval],
+                        early_stopping_rounds=5, verbose_eval=False)
+        self.assertEqual(gbm.best_iteration, 34)
+
+        def metrics_combination_train_regression(metric_list, assumed_iteration, first_metric_only, assertion=True):
+            params = {
+                'boosting_type': 'gbdt',
+                'objective': 'regression',
+                'learning_rate': 0.05,
+                'num_leaves': 5,
+                'metric': metric_list,
+                'verbose': -1,
+                'seed': 123,
+                'bagging_seed': 234,
+                'feature_fraction_seed': 456
+            }
+            gbm = lgb.train(dict(params, first_metric_only=first_metric_only), lgb_train,
+                            num_boost_round=300, valid_sets=[lgb_eval],
```

Comment:
`num_boost_round=300` times this function is called takes too much time! Please consider using <= 50 in the worst scenario. Maybe use another dataset or play around with metrics/early stopping rounds?..

Reply:
I understand your opinion, however if the num_boost_round is small, then the `best_iter` become same value among the metrics like `l1` and `l2`. In that case, I can not judge which metric is used on early stopping. So the `best_iter` should be different on these metrics. I will try to find best setting with small num_boost_round.

Reply:
And basically, these tests are applied early stopping before reaching num_boost_round, so I think  performance is not so worse.

Reply:
@matsuken92 Thanks! However, without concrete hardcoded value for `best_iter` assertion, there is no guarantee that `best_iter != num_boost_round` or `best_iter != num_boost_round - 1`. So, it's quite dangerous leave such big number of iterations.